### ISSUE TITLE
async. Increase requests from (1 each 5s) to (20 per second)

### ISSUE THREAD

**Issue by creslinux** [freqtrade/freqtrade#1046]
Currently the bot throttles its requests based on a value set in conf. 
This *appears to be to avoid API limits being hit, it impacts our response times and slows the entire bot as the loop is processed in serial. 

At the moment a pair may get processed once every 2-5 minutes depending if have a few open trades and 100 in the whitelist. We could get that delay down to many times  a second.

Binance supports returning all ticker for all markets in a single request 
I would propose 

      **Change FROM: calling each ticker individually, then processing 1 pair each request queuing all others**
      - e.g https://api.binance.com/api/v1/ticker/price?symbol=LTCBTC
      - limiting us to 1/req each 5 seconds if loading 120pairs
      
    **Change to: Requesting all markets each time and processing our entire whitelist loop each request**
      - e.g https://api.binance.com/api/v1/ticker/price
      - thereafter allowing FT to make 20 requests / second without hitting api limts

This will both allow more time accurate data, and speed up our loop potentially by 100x

Also binance returns its API limits in its market info response. 
 I would propose those values be put to use by default
 - the limits are at writing: 
    - 1200 ticker requests / minute
    - 10 orders  / second 
    - 100,000 orders  / day

**Comment by creslinux** [comment]
Sadly looks as though klines are not able to be requested on'mass as tickers are in single hit. 
This proposal a none starter before began

**Comment by creslinux** [comment]
For binance Klines (candle data) can only be requested per market. 

An ideal use of our 1,200 API requests per minute would be 
1  - On minute close download all market Klines (e.g up-to 130 requests, 1 each pair)
2 -  download all tickers for all markets in single request. 10 per second (600 requests each minute)

The leaves our API load comfortably under utilised at just over 60% (730 requests from 1200 / minute)

**Comment by vertti** [comment]
We should:
1) extract the candle history download out from get_signal
2) request all candle histories in one go using asyncio that is supported by CCXT ( https://github.com/ccxt/ccxt/wiki/Manual#synchronous-vs-asynchronous-calls) 
3) have the exchange class cache the ticker with a short time-to-live: when getting any ticker, download all tickers if cache is outdated, then return the requested ticker from the cache

**Comment by jblestang** [comment]
And maybe fetch the history the first time and then only update the latest candle?

**Comment by xmatthias** [comment]
i think that's covered in point 4 ...

**Comment by creslinux** [comment]
For reference here is a snippet i built that demonstrates getting all tickers in one hit vs a single

```
Example 1 ) Download only 1 ticker)
response = urllib.request.urlopen('https://api.binance.com/api/v1/ticker/price?symbol=LTCBTC')
ticker_pair = response.read()
print(ticker_pair)

Example 2) Download all tickers
response = urllib.request.urlopen('https://api.binance.com/api/v1/ticker/price')
tickers = json.loads(response.read())
#print(tickers)

for pair in tickers:
    if pair['symbol'] == 'LTCBTC':
        print(pair)
```

CCXT does support fetchTickers as a unified API also  request also. 

```
                             User
    +-------------------------------------------------------------+
    |                            CCXT                             |
    +------------------------------+------------------------------+
    |            Public            |           Private            |
    +=============================================================+
    │                              .                              |
    │                    The Unified CCXT API                     |
    │                              .                              |
    |       loadMarkets            .           fetchBalance       |
    |       fetchMarkets           .            createOrder       |
    |       fetchCurrencies        .            cancelOrder       |
    |       fetchTicker            .             fetchOrder       |
    |       fetchTickers           .            fetchOrders       |
    |       fetchOrderBook         .        fetchOpenOrders       |
    |       fetchOHLCV             .      fetchClosedOrders       |
    |       fetchTrades            .          fetchMyTrades       |
    |                              .                deposit       |
    |                              .               withdraw       |
    │                              .                              |
    +=============================================================+
```

**Comment by vertti** [comment]
@jblestang it's a good point and should probably be added as the 5th item. It's not really part of the 4. We should also measure is it faster. So that it's really worth the extra logic needed.

**Comment by jblestang** [comment]
I will be faster, less network data, less SSL decoding , and faster bot !

**Comment by creslinux** [comment]
Default is a depth of 500, max 1,000 both have an API weight of 1 
So there is no impact in terms of API cost. 

Saving in network time may be counter productive if an unusual request depth is not on the targets cache waiting for us to take a copy.  A 50byte payload will still be wrapped in a k of SSL/HTTP 

Id be tempted to take 500 in subsequent requests, its the most common and best chance we'll not run into unusual problems others have not worked through before for us.

**Comment by xmatthias** [comment]
i would doubt if the api servers of exchanges can implement caching effectivly (apart from keeping the last X candles in memory)
each request needs to be to the target (a caching proxy cannot apply for them, anyway) and needs to be invalidated as soon as a candle changes -and if that candle is half-full it changes per second ... ;) 

however, having the download concise and async should fix the network-latency ... the fewer candles is a "extra" which should not have high priority i think...

**Comment by creslinux** [comment]
I meant to say 500 is the default, its therefore the most likely to always be in their cloudflare cache.
It we request an artificially low candle count and miss the CND cache we'll be waiting for it to go all the way back to exchange before serving us. 

No large exchange is directly on the net, they're all behind CF to protect from DDOS.

**Comment by hroff-1902** [comment]
As far as I understand from the old comments above, this PR proposes to use ccxt `fetch_tickers()` method instead of currently fetched OHLCV for each pair, which should decrease the number of requests and allow thus to set throttling interval to be less than 1 sec. 

But, this will fetch "momental" tickers, and the candles should either be reconstructed on this data, or strategies (and most of the bot trading logic as well) to be reworked to be based on "momental" ticker data instead of the signals based on the last complete candle.

However, the use of fetch_tickers can be considered for other parts of the business logic, where "momental" data are okay -- like handling roi and stoplosses, these things that are checked at the middle of the candle...

**Comment by xmatthias** [comment]
The issue consists of 2 parts.
first, fetching OHLCV data in one go, once at the start of the next candle, async for all pairs in the whitelist - which is what has been implemented.

The second part proposes using `fetch_tickers()` once at iteration start and use the data to determine prices.
I do have a few concerns around this 2nd part. Depending on the strategy, analysis may take a second or 2 (assuming you use a few compute-intensive indicators).
* Assuming a pair generates a buy-signal, then we'd want the latest price, not a price we got 2 seconds ago - since the new price could be ... different.
*`fetch_tickers()` is a function (some?) exchanges document as "don't use for latest price, use orderbook instead".

For that reason, i think the 2nd part is better left as is (which would have us close this issue) - fetching the latest price right before creating the order - either from `fetch_ticker()` for that pair, or via the orderbook (recommended by exchanges) - to make sure we're using the latest prices.
